web/pages/api/v0/markets.ts

```diff
@@ -4,15 +4,46 @@ import { listAllContracts } from 'web/lib/firebase/contracts'
 import { applyCorsHeaders, CORS_UNRESTRICTED } from 'web/lib/api/cors'
 import { toLiteMarket } from './_types'
 
-type Data = any[]
-
 export default async function handler(
   req: NextApiRequest,
-  res: NextApiResponse<Data>
+  res: NextApiResponse
 ) {
   await applyCorsHeaders(req, res, CORS_UNRESTRICTED)
-  const contracts = await listAllContracts()
-  // Serve from Vercel cache, then update. see https://vercel.com/docs/concepts/functions/edge-caching
-  res.setHeader('Cache-Control', 's-maxage=1, stale-while-revalidate')
-  res.status(200).json(contracts.map(toLiteMarket))
+  let before: string | undefined
+  let n: number | undefined
+  if (req.query.before != null) {
+    if (typeof req.query.before !== 'string') {
+      res.status(400).json({ error: 'before must be null or a market ID.' })
+      return
+    }
+    before = req.query.before
+  }
+  if (req.query.n != null) {
+    if (typeof req.query.n !== 'string') {
+      res
+        .status(400)
+        .json({ error: 'n must be null or a number of markets to return.' })
+      return
+    }
+    n = parseInt(req.query.n)
+  } else {
+    n = 1000
+  }
+  if (n < 1 || n > 1000) {
+    res.status(400).json({ error: 'n must be between 1 and 1000.' })
+    return
+  }
+
+  try {
+    const contracts = await listAllContracts(n, before)
+    // Serve from Vercel cache, then update. see https://vercel.com/docs/concepts/functions/edge-caching
+    res.setHeader('Cache-Control', 's-maxage=1, stale-while-revalidate')
```

Comment:
Hm, do you know how the Vercel cache interacts with different query params? Does everything just magically work (eg `/markets?limit=8&before=blah` gets cached for 15 min)?

Reply:
I don't really know. Glancing at the documentation, it looks like probably each independent URL will be cached separately for 1 second (not sure where you got the 15 minute idea?)

In general, I think this whole endpoint is a bad idea. If people want a list of all markets, we should provide periodic database dumps. If people want to query markets programmatically, then an online API makes more sense. I don't think it makes sense to jam them both into the same code path and try to make the performance magically work out with clever caching headers.

Reply:
Oh, huh, I was mixing it up with the per-market endpoint which is updated every 2 minutes (not sure why I thought 15).

fwiw I'm on board with providing database dumps (maybe hourly?) for clients which want to go over everything and do analysis or something

Reply:
(Although we should probably make sure that our API at least returns client caching headers that make sense and works with etags. I have no idea whether it does.)